### PR TITLE
refactor: introduce primordials

### PR DESCRIPTION
**Pull request by lucacasonato**
This commit introduces primordials to deno_core. Primordials are a
frozen set of all intrinsic objects in the runtime. They are not
vulnerable to prototype pollution.

This commit just migrates deno_core to use primordials. Followup commits can
introduce primordials to the rest of the runtime.

Node has a eslint plugin that validates that all globals that are used in a
given script use primordials. We should consider adding a plugin for this to
dlint too. See https://github.com/nodejs/node/blob/master/tools/eslint-rules/prefer-primordials.js

Towards #10756

### CODE REVIEW

**Comment by benjamingr** [comment]
For what it's worth this is one of the ugliest things we have in the Node.js codebase and I'd caution against relying on primordials for security.

Please be very mindful when you consider taking this sort of burden upon yourself as a platform.

**Comment by lucacasonato** [comment]
@benjamingr I don't disagree, but there are some web platform tests we can not get to pass without using primordials or re-architechting the entire system to use only native v8 bindins. Example of a failing test: https://staging.wpt.fyi/results/fetch/api/response/response-stream-with-broken-then.any.html?product=deno&product=chrome.

The issue is that the glue between our native code ops our runtime internals are implemented as JS

We do not intend to use primordials for security - all our access control is and still will be managed by native code, and we do now and will continue to consider all JS code untrusted.

**Comment by benjamingr** [comment]
That's unfortunate, reasonable and fair.

As far as I know attempts to solve this (better) at the platform level were stopped - so I'm not sure what else you can do.

File: core/runtime.rs [hunk 1]
```diff
@@ -1732,7 +1732,8 @@ pub mod tests {
 
   #[test]
   fn test_heap_limits() {
-    let create_params = v8::Isolate::create_params().heap_limits(0, 20 * 1024);
+    let create_params =
+      v8::Isolate::create_params().heap_limits(0, 3 * 1024 * 1024);
```
Comment (lucacasonato) [inline core/runtime.rs:1736]: Note: required startup heap limits have increased.

Comment (bartlomieju) [reply core/runtime.rs:1736]: Increased a lot 😬

Comment (lucacasonato) [reply core/runtime.rs:1736]: V8 is really inefficient on first startup. You can see in the benchmarks that the snapshot (so baseline heap size of Deno) has not really increased.

**Comment by bartlomieju** [comment]
> Node has a eslint plugin that validates that all globals that are used in a
given script use primordials. We should consider adding a plugin for this to
dlint too.

@magurotuna how involved would be adding a lint rule like this to `dlint`?

**Comment by bartlomieju** [comment]
@bnoordhuis do you have any objections to this PR?

**Review by bartlomieju — APPROVED** [review]
LGTM

@lucacasonato could you later open an issue for tracking using primordials in extensions and `runtime/js` so we can share migration work?

**Review by piscisaureus — APPROVED** [review]
LGTM

**Comment by magurotuna** [comment]
@bartlomieju 

> how involved would be adding a lint rule like this to dlint?

Looks like rather complicated but not impossible. Let me try to implement it :)

**Review by bnoordhuis — APPROVED** [review]
LGTM and, for the record, I think it's a good idea.

**Comment by benjamingr** [comment]
> LGTM and, for the record, I think it's a good idea.

@bnoordhuis why?

**Comment by bnoordhuis** [comment]
@benjamingr Robustness in the presence of monkey patching. Same reason it was a good idea in Node. :-)

**Pull request merged**